DOCS/man/en/mplayer.1
.\" $Revision: 31552 $
.\" MPlayer (C) 2000-2010 MPlayer Team
.\" This man page was/is done by Gabucino, Diego Biurrun, Jonas Jermann
.
.\" --------------------------------------------------------------------------
.\" Macro definitions
.\" --------------------------------------------------------------------------
.
.\" define indentation for suboptions
.nr SS 5
.\" add new suboption
.de IPs
.IP "\\$1" \n(SS
..
.\" begin of first level suboptions, end with .RE
.de RSs
.RS 10
..
.\" begin of 2nd level suboptions
.de RSss
.PD 0
.RS \n(SS+3
..
.\" end of 2nd level suboptions
.de REss
.RE
.PD 1
..
.
.\" --------------------------------------------------------------------------
.\" Title
.\" --------------------------------------------------------------------------
.
.TH MPlayer 1 "2009-03-25" "The MPlayer Project" "The Movie Player"
.
.SH NAME
mplayer  \- movie player
.br
mencoder \- movie encoder
.
.\" --------------------------------------------------------------------------
.\" Synopsis
.\" --------------------------------------------------------------------------
.
.SH SYNOPSIS
.na
.nh
.B mplayer
[options] [file|URL|playlist|\-]
.
.br
.B mplayer
[options] file1
[specific options] [file2] [specific options]
.
.br
.B mplayer
[options]
{group of files and options}
[group-specific options]
.
.br
.B mplayer
[dvd|dvdnav]://[title|[start_title]\-end_title][/device]
[options]
.
.br
.B mplayer
vcd://track[/device]
.
.br
.B mplayer
tv://[channel][/input_id]
[options]
.
.br
.B mplayer
radio://[channel|frequency][/capture]
[options]
.
.br
.B mplayer
pvr://
[options]
.
.br
.B mplayer
dvb://[card_number@]channel
[options]
.
.br
.B mplayer
mf://[filemask|@listfile]
[\-mf options] [options]
.
.br
.B mplayer
[cdda|cddb]://track[\-endtrack][:speed][/device]
[options]
.
.br
.B mplayer
cue://file[:track]
[options]
.
.br
.B mplayer
[file|mms[t]|http|http_proxy|rt[s]p|ftp|udp|unsv|icyx|noicyx|smb]://
[user:pass@]URL[:port] [options]
.
.br
.B mplayer
sdp://file
[options]
.
.br
.B mplayer
mpst://host[:port]/URL
[options]
.
.br
.B mplayer
tivo://host/[list|llist|fsid]
[options]
.
.br
.B gmplayer
[options]
[\-skin\ skin]
.
.br
.B mencoder
[options] file
[file|URL|\-] [\-o file | file://file | smb://[user:pass@]host/filepath]
.
.br
.B mencoder
[options] file1
[specific options] [file2] [specific options]
.ad
.hy
.
.\" --------------------------------------------------------------------------
.\" Description
.\" --------------------------------------------------------------------------
.
.SH DESCRIPTION
.B mplayer
is a movie player for Linux (runs on many other platforms and CPU
architectures, see the documentation).
It plays most MPEG/\:VOB, AVI, ASF/\:WMA/\:WMV, RM, QT/\:MOV/\:MP4, Ogg/\:OGM,
MKV, VIVO, FLI, NuppelVideo, yuv4mpeg, FILM and RoQ files, supported by many
native and binary codecs.
You can watch VCD, SVCD, DVD, 3ivx, DivX 3/4/5, WMV and even H.264 movies,
too.
.PP
MPlayer supports a wide range of video and audio output drivers.
It works with X11, Xv, DGA, OpenGL, SVGAlib, fbdev, AAlib, libcaca, DirectFB,
Quartz, Mac OS X CoreVideo, but you can also use GGI, SDL (and all their drivers),
VESA (on every VESA-compatible card, even without X11), some low-level
card-specific drivers (for Matrox, 3dfx and ATI) and some hardware MPEG decoder
boards, such as the Siemens DVB, Hauppauge PVR (IVTV), DXR2 and DXR3/\:Hollywood+.
Most of them support software or hardware scaling, so you can enjoy movies in
fullscreen mode.
.PP
MPlayer has an onscreen display (OSD) for status information, nice big
antialiased shaded subtitles and visual feedback for keyboard controls.
European/\:ISO8859-1,2 (Hungarian, English, Czech, etc), Cyrillic and Korean
fonts are supported along with 12 subtitle formats (MicroDVD, SubRip, OGM,
SubViewer, Sami, VPlayer, RT, SSA, AQTitle, JACOsub, PJS and our own: MPsub) and
DVD subtitles (SPU streams, VOBsub and Closed Captions).
.PP
.B mencoder
(MPlayer's Movie Encoder) is a simple movie encoder, designed to encode
MPlayer-playable movies (see above) to other MPlayer-playable formats (see
below).
It encodes to MPEG-4 (DivX/Xvid), one of the libavcodec codecs and
PCM/\:MP3/\:VBRMP3 audio in 1, 2 or 3 passes.
Furthermore it has stream copying abilities, a powerful filter system (crop,
expand, flip, postprocess, rotate, scale, noise, RGB/\:YUV conversion) and
more.
.PP
.B gmplayer
is MPlayer with a graphical user interface.
It has the same options as MPlayer, however they might not all work correctly
due to conflicts with the configuration via the GUI (stored in gui.conf).
In particular some options might be overwritten by settings in gui.conf while
others might end up stored permanently in gui.conf.
.PP
Usage examples to get you started quickly can be found at the end
of this man page.
.PP
.B Also see the HTML documentation!
.
.
.\" --------------------------------------------------------------------------
.\" interactive control
.\" --------------------------------------------------------------------------
.
.SH "INTERACTIVE CONTROL"
MPlayer has a fully configurable, command-driven control layer
which allows you to control MPlayer using keyboard, mouse, joystick
or remote control (with LIRC).
See the \-input option for ways to customize it.
.
.TP
.B keyboard control
.PD 0
.RS
.IPs "<\- and \->"
Seek backward/\:forward 10 seconds.
.IPs "up and down"
Seek forward/\:backward 1 minute.
.IPs "pgup and pgdown"
Seek forward/\:backward 10 minutes.
.IPs "[ and ]"
Decrease/increase current playback speed by 10%.
.IPs "{ and }"
Halve/double current playback speed.
.IPs "backspace"
Reset playback speed to normal.
.IPs "< and >"
Go backward/\:forward in the playlist.
.IPs "ENTER"
Go forward in the playlist, even over the end.
.IPs "HOME and END"
next/\:previous playtree entry in the parent list
.IPs "INS and DEL (ASX playlist only)"
next/\:previous alternative source.
.IPs "p / SPACE"
Pause (pressing again unpauses).
.IPs ".\ \ \ \ "
Step forward.
Pressing once will pause movie, every consecutive press will play one frame
and then go into pause mode again (any other key unpauses).
.IPs "q / ESC"
Stop playing and quit.
.IPs "U\ \ \ \ "
Stop playing (and quit if \-idle is not used).
.IPs "+ and \-"
Adjust audio delay by +/\- 0.1 seconds.
.IPs "/ and *"
Decrease/\:increase volume.
.IPs "9 and 0"
Decrease/\:increase volume.
.IPs "( and )"
Adjust audio balance in favor of left/\:right channel.
.IPs "m\ \ \ \ "
Mute sound.
.IPs "_ (MPEG-TS, AVI and libavformat only)"
Cycle through the available video tracks.
.IPs "# (DVD, MPEG, Matroska, AVI and libavformat only)"
Cycle through the available audio tracks.
.IPs "TAB (MPEG-TS and libavformat only)"
Cycle through the available programs.
.IPs "f\ \ \ \ "
Toggle fullscreen (also see \-fs).
.IPs "T\ \ \ \ "
Toggle stay-on-top (also see \-ontop).
.IPs "w and e"
Decrease/\:increase pan-and-scan range.
.IPs "o\ \ \ \ "
Toggle OSD states: none / seek / seek + timer / seek + timer + total time.
.IPs "d\ \ \ \ "
Toggle frame dropping states: none / skip display / skip decoding
(see \-framedrop and \-hardframedrop).
.IPs "v\ \ \ \ "
Toggle subtitle visibility.
.IPs "j\ \ \ \ "
Cycle through the available subtitles.
.IPs "y and g"
Step forward/backward in the subtitle list.
.IPs "F\ \ \ \ "
Toggle displaying "forced subtitles".
.IPs "a\ \ \ \ "
Toggle subtitle alignment: top / middle / bottom.
.IPs "x and z"
Adjust subtitle delay by +/\- 0.1 seconds.
.IPs "r and t"
Move subtitles up/down.
.IPs "i (\-edlout mode only)"
Set start or end of an EDL skip and write it out to the given file.
.IPs "s (\-vf screenshot only)"
Take a screenshot.
.IPs "S (\-vf screenshot only)"
Start/stop taking screenshots.
.IPs "I\ \ \ \ "
Show filename on the OSD.
.IPs "P\ \ \ \ "
Show progression bar, elapsed time and total duration on the OSD.
.IPs "! and @"
Seek to the beginning of the previous/next chapter.
.IPs "D (\-vo xvmc, \-vo vdpau, \-vf yadif, \-vf kerndeint only)"
Activate/deactivate deinterlacer.
.IPs "A"
Cycle through the available DVD angles.
.RE
.PD 1
.PP
.RS
(The following keys are valid only when using a hardware accelerated video
output (xv, (x)vidix, (x)mga, etc), the software equalizer
(\-vf eq or \-vf eq2) or hue filter (\-vf hue).)
.RE
.PP
.PD 0
.RS
.IPs "1 and 2"
Adjust contrast.
.IPs "3 and 4"
Adjust brightness.
.IPs "5 and 6"
Adjust hue.
.IPs "7 and 8"
Adjust saturation.
.RE
.PD 1
.PP
.RS
(The following keys are valid only when using the quartz or corevideo
video output driver.)
.RE
.PP
.PD 0
.RS
.IPs "command + 0"
Resize movie window to half its original size.
.IPs "command + 1"
Resize movie window to its original size.
.IPs "command + 2"
Resize movie window to double its original size.
.IPs "command + f"
Toggle fullscreen (also see \-fs).
.IPs "command + [ and command + ]"
Set movie window alpha.
.RE
.PD 1
.PP
.RS
(The following keys are valid only when using the sdl
video output driver.)
.RE
.PP
.PD 0
.RS
.IPs "c\ \ \ \ "
Cycle through available fullscreen modes.
.IPs "n\ \ \ \ "
Restore original mode.
.RE
.PD 1
.PP
.RS
(The following keys are valid if you have a keyboard
with multimedia keys.)
.RE
.PP
.PD 0
.RS
.IPs PAUSE
Pause.
.IPs "STOP\ "
Stop playing and quit.
.IPs "PREVIOUS and NEXT"
Seek backward/\:forward 1 minute.
.RE
.PD 1
.PP
.RS
(The following keys are only valid if GUI support is compiled in
and will take precedence over the keys defined above.)
.RE
.PP
.PD 0
.RS
.IPs ENTER
Start playing.
.IPs "ESC\ \ "
Stop playing.
.IPs "l\ \ \ \ "
Load file.
.IPs "t\ \ \ \ "
Load subtitle.
.IPs "c\ \ \ \ "
Open skin browser.
.IPs "p\ \ \ \ "
Open playlist.
.IPs "r\ \ \ \ "
Open preferences.
.RE
.PD 1
.PP
.RS
(The following keys are only valid if you compiled with TV or DVB input
support and will take precedence over the keys defined above.)
.RE
.PP
.PD 0
.RS
.IPs "h and k"
Select previous/\:next channel.
.IPs "n\ \ \ \ "
Change norm.
.IPs "u\ \ \ \ "
Change channel list.
.RE
.PD 1
.PP
.RS
(The following keys are only valid if you compiled with dvdnav
support: They are used to navigate the menus.)
.RE
.PP
.PD 0
.RS
.IPs "keypad 8"
Select button up.
.IPs "keypad 2"
Select button down.
.IPs "keypad 4"
Select button left.
.IPs "keypad 6"
Select button right.
.IPs "keypad 5"
Return to main menu.
.IPs "keypad 7"
Return to nearest menu (the order of preference is: chapter->title->root).
.IPs "keypad ENTER"
Confirm choice.
.RE
.PD 1
.PP
.RS
.
(The following keys are only valid if teletext support is enabled during
compilation: They are used for controlling TV teletext, whose data may come
from either an analog TV source or an MPEG Transport Stream.)
.RE
.PP
.PD 0
.RS
.IPs "X\ \ \ \ "
Switch teletext on/\:off.
.IPs "Q and W"
Go to next/\:prev teletext page.
.RE
.PD 1
.
.TP
.B mouse control
.PD 0
.RS
.IPs "button 3 and button 4"
Seek backward/\:forward 1 minute.
.IPs "button 5 and button 6"
Decrease/\:increase volume.
.RE
.PD 1
.
.TP
.B joystick control
.PD 0
.RS
.IPs "left and right"
Seek backward/\:forward 10 seconds.
.IPs "up and down"
Seek forward/\:backward 1 minute.
.IPs "button 1"
Pause.
.IPs "button 2"
Toggle OSD states: none / seek / seek + timer / seek + timer + total time.
.IPs "button 3 and button 4"
Decrease/\:increase volume.
.RE
.PD 1
.
.
.\" --------------------------------------------------------------------------
.\" Usage
.\" --------------------------------------------------------------------------
.
.SH "USAGE"
Every 'flag' option has a 'noflag' counterpart, e.g.\& the opposite of the
\-fs option is \-nofs.
.PP
If an option is marked as (XXX only), it will only work in combination with
the XXX option or if XXX is compiled in.
.PP
.I NOTE:
The suboption parser (used for example for \-ao pcm suboptions) supports
a special kind of string-escaping intended for use with external GUIs.
.br
It has the following format:
.br
%n%string_of_length_n
.br
.I EXAMPLES:
.br
mplayer \-ao pcm:file=%10%C:test.wav test.avi
.br
Or in a script:
.br
mplayer \-ao pcm:file=%`expr length "$NAME"`%"$NAME" test.avi
.PP
.
.\" --------------------------------------------------------------------------
.\" Configuration files
.\" --------------------------------------------------------------------------
.
.SH "CONFIGURATION FILES"
You can put all of the options in configuration files which will be read
every time MPlayer/MEncoder is run.
The system-wide configuration file 'mplayer.conf' is in your configuration
directory (e.g.\& /etc/\:mplayer or /usr/\:local/\:etc/\:mplayer), the user
specific one is '~/\:.mplayer/\:config'.
The configuration file for MEncoder is 'mencoder.conf' in your configuration
directory (e.g.\& /etc/\:mplayer or /usr/\:local/\:etc/\:mplayer), the
user specific one is '~/\:.mplayer/\:mencoder.conf'.
User specific options override system-wide options and options given on the
command line override either.
The syntax of the configuration files is 'option=<value>', everything after
a '#' is considered a comment.
Options that work without values can be enabled by setting them to 'yes'
or '1' or 'true' and disabled by setting them to 'no' or '0' or 'false'.
Even suboptions can be specified in this way.
.PP
You can also write file-specific configuration files.
If you wish to have a configuration file for a file called 'movie.avi', create a file
named 'movie.avi.conf' with the file-specific options in it and put it in
~/.mplayer/.
You can also put the configuration file in the same directory as the file to
be played, as long as you give the \-use\-filedir\-conf option (either on the
command line or in your global config file).
If a file-specific configuration file is found in the same directory, no
file-specific configuration is loaded from ~/.mplayer.
In addition, the \-use\-filedir\-conf option enables directory-specific
configuration files.
For this, MPlayer first tries to load a mplayer.conf from the same directory as
the file played and then tries to load any file-specific configuration.
.PP
.I EXAMPLE MPLAYER CONFIGURATION FILE:
.sp 1
.nf
# Use Matrox driver by default.
vo=xmga
# I love practicing handstands while watching videos.
flip=yes
# Decode/encode multiple files from PNG,
# start with mf://filemask
mf=type=png:fps=25
# Eerie negative images are cool.
vf=eq2=1.0:\-0.8
.fi
.PP
.I "EXAMPLE MENCODER CONFIGURATION FILE:"
.sp 1
.nf
# Make MEncoder output to a default filename.
o=encoded.avi
# The next 4 lines allow mencoder tv:// to start capturing immediately.
oac=pcm=yes
ovc=lavc=yes
lavcopts=vcodec=mjpeg
tv=driver=v4l2:input=1:width=768:height=576:device=/dev/video0:audiorate=48000
# more complex default encoding option set
lavcopts=vcodec=mpeg4:autoaspect=1
lameopts=aq=2:vbr=4
ovc=lavc=1
oac=lavc=1
passlogfile=pass1stats.log
noautoexpand=1
subfont-autoscale=3
subfont-osd-scale=6
subfont-text-scale=4
subalign=2
subpos=96
spuaa=20
.fi
.
.\" --------------------------------------------------------------------------
.\" Profiles
.\" --------------------------------------------------------------------------
.
.SH "PROFILES"
To ease working with different configurations profiles can be defined in the
configuration files.
A profile starts with its name between square brackets, e.g.\& '[my-profile]'.
All following options will be part of the profile.
A description (shown by \-profile help) can be defined with the profile-desc
option.
To end the profile, start another one or use the profile name 'default'
to continue with normal options.
.fi
.PP
.I "EXAMPLE MPLAYER PROFILE:"
.sp 1
.nf

[protocol.dvd]
profile-desc="profile for dvd:// streams"
vf=pp=hb/vb/dr/al/fd
alang=en

[protocol.dvdnav]
profile-desc="profile for dvdnav:// streams"
profile=protocol.dvd
mouse-movements=yes
nocache=yes

[extension.flv]
profile-desc="profile for .flv files"
flip=yes

[vo.pnm]
outdir=/tmp

[ao.alsa]
device=spdif
.fi
.PP
.I "EXAMPLE MENCODER PROFILE:"
.sp 1
.nf

[mpeg4]
profile-desc="MPEG4 encoding"
ovc=lacv=yes
lavcopts=vcodec=mpeg4:vbitrate=1200

[mpeg4\-hq]
profile-desc="HQ MPEG4 encoding"
profile=mpeg4
lavcopts=mbd=2:trell=yes:v4mv=yes
.fi
.
.\" --------------------------------------------------------------------------
.\" Options
.\" --------------------------------------------------------------------------
.
.SH "GENERAL OPTIONS"
.
.TP
.B \-codecpath <dir>
Specify a directory for binary codecs.
.
.TP
.B \-codecs\-file <filename> (also see \-afm, \-ac, \-vfm, \-vc)
Override the standard search path and use the specified file
instead of the builtin codecs.conf.
.
.TP
.B \-include <configuration file>
Specify configuration file to be parsed after the default ones.
.
.TP
.B \-list\-options
Prints all available options.
.
.TP
.B \-msgcharset <charset>
Convert console messages to the specified character set (default: autodetect).
Text will be in the encoding specified with the \-\-charset configure option.
Set this to "noconv" to disable conversion (for e.g.\& iconv problems).
.br
.I NOTE:
The option takes effect after command line parsing has finished.
The MPLAYER_CHARSET environment variable can help you get rid of
the first lines of garbled output.
.
.TP
.B \-msgcolor
Enable colorful console output on terminals that support ANSI color.
.
.TP
.B \-msglevel <all=<level>:<module>=<level>:...>
Control verbosity directly for each module.
The 'all' module changes the verbosity of all the modules not
explicitly specified on the command line.
See '\-msglevel help' for a list of all modules.
.br
.I NOTE:
Some messages are printed before the command line is parsed and are
therefore not affected by \-msglevel.
To control these messages you have to use the MPLAYER_VERBOSE environment
variable, see its description below for details.
.br
Available levels:
.PD 0
.RSs
.IPs "\-1"
complete silence
.IPs " 0"
fatal messages only
.IPs " 1"
error messages
.IPs " 2"
warning messages
.IPs " 3"
short hints
.IPs " 4"
informational messages
.IPs " 5"
status messages (default)
.IPs " 6"
verbose messages
.IPs " 7"
debug level 2
.IPs " 8"
debug level 3
.IPs " 9"
debug level 4
.RE
.PD 1
.
.TP
.B \-msgmodule
Prepend module name in front of each console message.
.
.TP
.B \-noconfig <options>
Do not parse selected configuration files.
.br
.I NOTE:
If \-include or \-use\-filedir\-conf options are
specified at the command line, they will be honoured.
.sp 1
Available options are:
.PD 0
.RSs
.IPs "all\ \ "
all configuration files
.IPs "gui (GUI only)"
GUI configuration file
.IPs system
system configuration file
.IPs "user\ "
user configuration file
.RE
.PD 1
.
.TP
.B "\-quiet\ "
Make console output less verbose; in particular, prevents the status line
(i.e.\& A:   0.7 V:   0.6 A-V:  0.068 ...) from being displayed.
Particularly useful on slow terminals or broken ones which do not properly
handle carriage return (i.e.\& \\r).
.
.TP
.B \-priority <prio> (Windows and OS/2 only)
Set process priority for MPlayer according to the predefined
priorities available under Windows and OS/2.
Possible values of <prio>:
.RSs
idle|belownormal|normal|abovenormal|high|realtime
.RE
.sp 1
.RS
.I WARNING:
Using realtime priority can cause system lockup.
.RE
.
.TP
.B \-profile <profile1,profile2,...>
Use the given profile(s), \-profile help displays a list of the defined profiles.
.
.TP
.B \-really\-quiet (also see \-quiet)
Display even less output and status messages than with \-quiet.
Also suppresses the GUI error message boxes.
.
.TP
.B \-show\-profile <profile>
Show the description and content of a profile.
.
.TP
.B \-use\-filedir\-conf
Look for a file-specific configuration file in the same directory as
the file that is being played.
.br
.I WARNING:
May be dangerous if playing from untrusted media.
.
.TP
.B "\-v\ \ \ \ \ "

MPlayer.cpp
/* started from VorbisPlayer
 *
 * okkwon -_-;
 *
 */

#define LOG_TAG "MPlayer"
#include "utils/Log.h"

#include <stdio.h>
#include <assert.h>
#include <limits.h>
#include <unistd.h>
#include <fcntl.h>
#include <sched.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <ui/GraphicBuffer.h>
#include "mplayer_lib.h"
#include "MPlayer.h"
#include "MPlayerRenderer.h"
#include "MPDebug.h"

#ifdef HAVE_GETTID
static pid_t myTid() { return gettid(); }
#else
static pid_t myTid() { return getpid(); }
#endif

namespace android {
	static status_t ERROR_NOT_OPEN = -1;
	static status_t ERROR_OPEN_FAILED = -2;
	static status_t ERROR_ALLOCATE_FAILED = -4;
	static status_t ERROR_NOT_SUPPORTED = -8;
	static status_t ERROR_NOT_READY = -16;
	static status_t STATE_INIT = 0;
	static status_t STATE_ERROR = 1;
	static status_t STATE_OPEN = 2;
	static status_t STATE_PREPARED = 3;


	bool MPlayer::libInUse = false;

	MPlayer::MPlayer() :
		mAudioBuffer(NULL), mState(STATE_ERROR),
		mLoop(false), mAndroidLoop(false),
		mExit(false), mPaused(false), mRender(false), mRenderTid(-1),
		mMPInitialized(false), mVideoRenderer(NULL)
	{
		LOGE("constructor\n");
	}

	void MPlayer::onFirstRef()
	{
		LOGE("initCheck\n");
		Mutex::Autolock l(mMutex);
		createThreadEtc(renderThread, this, "mplayer main loop", 
				ANDROID_PRIORITY_AUDIO);
		mCondition.wait (mMutex);
		if (mRenderTid > 0) {
			LOGV("render thread(%d) started", mRenderTid);
			mState = STATE_INIT;
		}

		/* todo : need to decide this -_-? */
		mDisplayWidth = 1024;
		mDisplayHeight = 590;
	}

	status_t MPlayer ::initCheck()
	{
		if (mState == STATE_ERROR) return ERROR_NOT_READY;
		return NO_ERROR;
	}

	MPlayer::~MPlayer() {
	    LOGE("VorbisPlayer destructor\n");
		release();
	}

	status_t MPlayer::setDataSource(const char* path)
	{
		return setdatasource(path, -1, 0, 0x78ffffLL);
	}

	status_t MPlayer::setDataSource(int fd, int64_t offset, int64_t length)
	{
		return setdatasource(NULL, fd, offset, length);
	}

	status_t MPlayer::setdatasource (const char *path, int fd, int64_t offset,
			int64_t length)
	{
		int ret;

		LOGE("setdatasource url%d, fd%d\n", path, fd);
		LOGE("setdatasource offset%d, length%d\n", offset, length);

		if (libInUse) return INVALID_OPERATION;

		Mutex::Autolock l(mMutex);
		if (mState == STATE_OPEN) {
			reset_nosync();
		}

		struct stat sb;
		if (path) {
			ret = stat(path, &sb);
		} else {
			ret = fstat(fd, &sb);
		}
		if (ret != 0) {
			mState = STATE_ERROR;
			close(fd);
			return ERROR_OPEN_FAILED;
		}
		/* todo : need to handle offset and length */

		int argca;
		const char * argva[30] = {"mplayer", "fd", 
			"-screenw", "1024", "-screenh", "550",
			"-vf", "dsize=x:y:0,scale",
			"-vo", "mem", "-framedrop",
			"-ao", "pcm_mem", "-noconsolecontrols", "-nojoystick",
			"-nolirc", "-nomouseinput", "-slave", "-zoom", "-fs",
			0};
		char url_buffer[100];
		char screenh_buffer[30];
		char screenw_buffer[30];
		char scale_buffer[30];

		int mfd = dup(fd);

		sprintf (url_buffer, "fd://%d", mfd);
		argva[1] = url_buffer;
		sprintf (screenw_buffer, "%d", mDisplayWidth);
		argva[3] = screenw_buffer;
		sprintf (screenh_buffer, "%d", mDisplayHeight);
		argva[5] = screenh_buffer;
		sprintf (scale_buffer, "dsize=%d:%d:0,scale",
			   	mDisplayWidth, mDisplayHeight);
		argva[7] = scale_buffer;

		for (argca=0; argca<30; argca++) {
			if (argva[argca] == 0)
				break;
		}
		libInUse = true;
		ret = mplayer_init (&mMPContext, argca, argva);
		ret = fstat(mfd, &sb);
		LOGV("okkwon stat result %d", ret);

		if (!ret) { 
			libInUse = true;
			mMPInitialized = true;
			mState = STATE_OPEN;
			return NO_ERROR;
		} else {
			mplayer_close (&mMPContext);
			libInUse = false;
			return ERROR_OPEN_FAILED;
		}
	}

	void MPlayer::depopulateISurface() {
		if (mVideoRenderer) {
			delete mVideoRenderer;
			mVideoRenderer = NULL;
		}
	}

	void MPlayer::populateISurface() {
		if (mState != STATE_OPEN) {
			return;
		}

		mVideoRenderer = new MPlayerRenderer (mISurface, mDisplayWidth,
				mDisplayHeight);
		/*
		if (mVideoRenderer) {
			int ret;
			int width, height;
			ret = mplayer_get_video_size(&mMPContext, &width, &height);
			if (ret) {
				mVideoRenderer->getVideoOutSize (width, height,
						&mVideoOutWidth, &mVideoOutHeight);
				LOGE("video size w%d h%d", mVideoOutWidth, mVideoOutHeight);
			}
		}
		*/
	}

	status_t MPlayer::setVideoSurface(const sp<ISurface> &isurface) {
		LOGE("setVideoSurface");

		Mutex::Autolock l(mMutex);
		if (mState != STATE_INIT && mState != STATE_OPEN) {
			return NO_INIT;
		}

		depopulateISurface();

		mISurface = isurface;

		return NO_ERROR;
	}


	status_t MPlayer::prepare()
	{
		LOGE("prepare\n");
		if (mState != STATE_OPEN) {
			return ERROR_NOT_OPEN;
		}
		return NO_ERROR;
	}

	status_t MPlayer::prepareAsync()
	{
		LOGE("prepareAsync\n");
		
		mplayer_prepare_play(&mMPContext);
		if (mISurface.get() != NULL) {
			populateISurface();
		}

		if (mState != STATE_OPEN) {
			sendEvent (MEDIA_ERROR);
			return NO_ERROR;
		}


		sendEvent (MEDIA_PREPARED);	/* todo : should be moved to main loop */
		return NO_ERROR;
	}

	status_t MPlayer::start()
	{
		LOGE("start\n");
		Mutex::Autolock l(mMutex);
		if (mState != STATE_OPEN) {
			return ERROR_NOT_OPEN;
		}
		mPaused = false;
		mRender = true;

		// wake up render thread
		LOGE("  wakeup render thread\n");
		mCondition.signal();
		return NO_ERROR;
	}

	status_t MPlayer::stop()
	{
		LOGE("stop\n");
		Mutex::Autolock l(mMutex);
		if (mState != STATE_OPEN) {
			return ERROR_NOT_OPEN;
		}
		mPaused = true;
		mRender = false;
		return NO_ERROR;
	}

	status_t MPlayer::seekTo(int position)
	{
		LOGE("seekTo pos%d\n", position);
		Mutex::Autolock l(mMutex);
		if (mState != STATE_OPEN) {
			return ERROR_NOT_OPEN;
		}
		mplayer_seek (&mMPContext, position);
		sendEvent(MEDIA_SEEK_COMPLETE);
		return NO_ERROR;
	}

	status_t MPlayer::pause()
	{
		LOGE("pause\n");
		Mutex::Autolock l(mMutex);
		if (mState != STATE_OPEN) {
			return ERROR_NOT_OPEN;
		}
		mPaused = true;
		return NO_ERROR;
	}

	bool MPlayer::isPlaying()
	{
		LOGE("isPlaying\n");
		if (mState == STATE_OPEN) {
			return mRender;
		}
		return 1;
	}

	status_t MPlayer::getCurrentPosition(int* position)
	{
		LOGE("getCurrentPosition\n");
		Mutex::Autolock l(mMutex);
		if (mState != STATE_OPEN) {
			LOGE("getCurrentPosition(): file not open");
			return ERROR_NOT_OPEN;
		}
		/* todo : get the position */
		mplayer_get_pos (&mMPContext, position);
		LOGE("position %d", *position);
		return NO_ERROR;
	}

	status_t MPlayer::getDuration(int *duration)
	{
		LOGE("getDuration\n");
		Mutex::Autolock l(mMutex);
		if (mState != STATE_OPEN) {
			return ERROR_NOT_OPEN;
		}
		mplayer_get_duration (&mMPContext, duration);
		LOGE("duration %d", *duration);
		return NO_ERROR;
	}

	status_t MPlayer::release()
	{
		LOGE("release\n");
		Mutex::Autolock l(mMutex);
		reset_nosync();

		if (mRenderTid > 0) {
			mExit = true;
			mCondition.signal();
			mCondition.wait(mMutex);
		}
		return NO_ERROR;
	}

	status_t MPlayer::reset()
	{
		LOGE("reset\n");
		Mutex::Autolock l(mMutex);

		return reset_nosync();
	}

	status_t MPlayer::reset_nosync()
	{
		//close file
		if (mMPInitialized)
		   	mplayer_close(&mMPContext);
		mMPInitialized = false;
		mState = STATE_ERROR;
		libInUse = false;

		mLoop = false;
		mAndroidLoop = false;
		mPaused = false;
		mRender = false;
		return NO_ERROR;
	}

	status_t MPlayer::setLooping(int loop)
	{
		LOGE("setLooping %d\n", loop);
		Mutex::Autolock l(mMutex);
		mLoop = (loop != 0);
		return NO_ERROR;
	}

	status_t MPlayer::createOutputTrack() {
		uint32_t sampleRate;
	   	int channelCount;
		
		mplayer_get_audio_info(&mMPContext, &sampleRate, &channelCount);

		if (mAudioSink->open(sampleRate, channelCount,
				   	AudioSystem::PCM_16_BIT,
				   	DEFAULT_AUDIOSINK_BUFFERCOUNT) != NO_ERROR) {
			LOGE("mAudioSink open failed");
			return ERROR_OPEN_FAILED;
		}
		return NO_ERROR;
	}

	int MPlayer::renderThread(void*p) {
		return ((MPlayer*)p)->render();
	}

#define AUDIOBUFFER_SIZE (4096*10)

	int MPlayer::render() {
		int result = -1;
		int temp;
		bool audioStarted = false;
		int mpresult;
		int audio_pos;
		char * video_buffer;
		size_t video_buffer_size;
		
		mAudioBuffer = new char[AUDIOBUFFER_SIZE];

		// let main thread know we're ready
		{
		   	Mutex::Autolock l(mMutex);
			mRenderTid = myTid();
			mCondition.signal();
		}

		while (1) {
			int audio_pos;
			int decoded_frames;

			Mutex::Autolock l(mMutex);

			// pausing?
			if (mPaused) {
				if (mAudioSink->ready()) mAudioSink->pause();
				mRender = false;
				audioStarted = false;
			}

			//nothing to render?
			if (!mExit && !mRender) {
				LOGE("render - signal wait\n");
				mCondition.wait(mMutex);
				LOGE("render - signal rx\n");
			}
			if (mExit) break;

			if (!mRender) continue;

			mpresult = mplayer_decode_audio(&mMPContext, mAudioBuffer,
					AUDIOBUFFER_SIZE, &audio_pos);
			if (mpresult) {
				LOGE ("error in decode");
				sendEvent(MEDIA_ERROR);
				break;
			}

			//create audio output track
			if (!mAudioSink->ready()) {
				LOGE("create output track");
				if (createOutputTrack() != NO_ERROR) break;
			}

			//write data to the audio hardware
			if ((temp = mAudioSink->write(mAudioBuffer, audio_pos))<0) {
				LOGE("Error in writing %d", temp);
				result = temp;
				break;
			}

			//start audio output if necessary
			if (!audioStarted && !mPaused && !mExit) {
				mAudioSink->start();
				audioStarted = true;
			}

			if (mVideoRenderer) {
				mVideoRenderer->getBuffer (&video_buffer, &video_buffer_size);
			}

			//this function will decode video and sleep for video output timing
			mpresult |= mplayer_decode_video(&mMPContext, video_buffer, 
					&decoded_frames);
			if (mpresult)  {
				LOGE("mplayer_decode_video returned %d", mpresult);
				break;
			}

			if (mVideoRenderer && decoded_frames) {
				mVideoRenderer->renderBuffer();
			}

			mpresult |= mplayer_after_decode (&mMPContext);
			if (mpresult) {
				LOGE("mplayer_after_decode %d", mpresult);
				break;
			}

			if (mpresult) {
				/* EOF reached -_-? */
				if (mLoop || mAndroidLoop) {
					mplayer_seek (&mMPContext, 0);
					/*
					mpresult = mplayer_decode_audio(&mMPContext, mAudioBuffer,
							AUDIOBUFFER_SIZE, &audio_pos);
							*/
				} else {
					int endpos;

					mAudioSink->stop();
					audioStarted = false;
					mRender = false;
					mPaused = true;
					mplayer_get_pos (&mMPContext, &endpos);

					LOGE("send MEDIA_PLAYBACK_COMPLETE");
					sendEvent(MEDIA_PLAYBACK_COMPLETE);

					LOGE("playback complete - wait for signal");
					mCondition.wait(mMutex);
					LOGE("playback complete - signal rx'd");
					if (mExit) break;

					if (mState == STATE_OPEN) {
						int curpos;
						mplayer_get_pos(&mMPContext, &curpos);
						if (curpos == endpos) {
							mplayer_seek (&mMPContext, 0);
						}
						mplayer_decode_audio(&mMPContext, mAudioBuffer,
								AUDIOBUFFER_SIZE, &audio_pos);
					}
				}
			}

		}
threadExit:
		mAudioSink.clear();
		if (mAudioBuffer) {
			delete [] mAudioBuffer;
			mAudioBuffer = NULL;
		}

		//tell main thread goodbye
		Mutex::Autolock l(mMutex);
		mRenderTid = -1;
		mCondition.signal();
		return result;
	}
} //end namespace android
